tests/crypto/src/common_test.c

```diff
@@ -0,0 +1,195 @@
+#include <drivers/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+
+#include "common_test.h"
+
+mbedtls_ctr_drbg_context ctr_drbg_ctx;
+
+static int entropy_func(void *ctx, unsigned char *buf, size_t len)
+{
+	return entropy_get_entropy(ctx, buf, len);
+}
+
+int init_ctr_drbg(const unsigned char *p_optional_seed, size_t len)
+{
+	static const unsigned char ncs_seed[] = "ncs_drbg_seed";
+
+	const unsigned char *p_seed;
+
+	if (p_optional_seed == NULL) {
+		p_seed = ncs_seed;
+		len = sizeof(ncs_seed);
+	} else {
+		p_seed = p_optional_seed;
+	}
+
+	struct device *p_device = device_get_binding(CONFIG_ENTROPY_NAME);
+	if (!p_device) {
+		return -ENODEV;
+	}
+
+	mbedtls_ctr_drbg_init(&ctr_drbg_ctx);
+	return mbedtls_ctr_drbg_seed(&ctr_drbg_ctx, entropy_func, p_device, p_seed, len);
+}
+
+uint32_t unhexify(uint8_t *p_output, char const *p_input)
+{
+	unsigned char c, c2;
+
+	if (p_input == NULL) {
+		return 0;
+	}
+
+	if (p_output == NULL) {
+		return 0;
+	}
+
+	int len = strlen(p_input) / 2;
+
+	while (*p_input != 0) {
+		c = *p_input++;
+		if (c >= '0' && c <= '9') {
+			c -= '0';
+		} else if (c >= 'a' && c <= 'f') {
+			c -= 'a' - 10;
+		} else if (c >= 'A' && c <= 'F') {
+			c -= 'A' - 10;
+		} else   {
+			printk("unhexify ERROR");
+		}
+
+		c2 = *p_input++;
+		if (c2 >= '0' && c2 <= '9') {
+			c2 -= '0';
+		} else if (c2 >= 'a' && c2 <= 'f') {
+			c2 -= 'a' - 10;
+		} else if (c2 >= 'A' && c2 <= 'F') {
+			c2 -= 'A' - 10;
+		} else   {
+			printk("unhexify ERROR");
+		}
+
+		*p_output++ = (c << 4) | c2;
+	}
+
+	return len;
+}
+
+const char *get_vector_name(const test_case_t *tc, uint32_t v)
+{
+	switch (tc->vector_type) {
+	case VT_HASH:
+	{
+		test_vector_hash_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_HMAC:
+	{
+		test_vector_hmac_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_HKDF:
+	{
+		test_vector_hkdf_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_ECDH:
+	{
+		test_vector_ecdh_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_ECDSA_SIGN:
+	{
+		test_vector_ecdsa_sign_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_ECDSA_VERIFY:
+	{
+		test_vector_ecdsa_verify_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_ECDSA_RANDOM:
+	{
+		test_vector_ecdsa_random_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_AES:
+	{
+		test_vector_aes_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_AEAD:
+	{
+		test_vector_aead_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+	case VT_ECJPAKE:
+	{
+		test_vector_ecjpake_t *tv = tc->vectors_start;
+		return tv[v].p_test_vector_name;
+	}
+
+	default:
+		return 0;
+	}
+}
+
+uint32_t get_vector_count(const test_case_t *tc)
+{
+	switch (tc->vector_type) {
+	case VT_HASH:
+		return ((size_t)tc->vectors_stop - (size_t)tc->vectors_start) / sizeof(test_vector_hash_t);
```

Comment:
Put these sizes in an array, and then do:

`return ((size_t)tc->vectors_stop - (size_t)tc->vectors_start) / test_case_vector_sizes[tc->vector_type];`

See if you can do similar refactoring elsewhere as well.

Reply:
Done in 1ce8d33